core/src/main/scala/tamer/Tamer.scala

```diff
@@ -81,76 +148,56 @@ object Tamer {
-      val partitionOffsets = stateConsumer.endOffsets(partitionSet)
-      val committedPartitionOffsets = stateConsumer.committed(partitionSet).map {
-        _.map {
-          case (tp, Some(o)) => Some((tp, o.offset()))
-          case _             => None
-        }.flatten.toMap
-      }
-      partitionOffsets.zip(committedPartitionOffsets).map {
-        case (po, _) if po.values.forall(_ == 0L) => Initialize
-        case Lag(lags) if lags.values.sum == 1L   => Resume
-        case Lag(lags) if lags.values.sum == 0L   => Retry
-        case _                                    => Fail
+    val decide: Set[TopicPartition] => Task[Decision] =
+      (partitionSet: Set[TopicPartition]) => {
+        val partitionOffsets = stateConsumer.endOffsets(partitionSet)
+        val committedPartitionOffsets = stateConsumer.committed(partitionSet).map {
+          _.map {
+            case (tp, Some(o)) => Some((tp, o.offset()))
+            case _             => None
+          }.flatten.toMap
+        }
+        partitionOffsets.zip(committedPartitionOffsets).map { case (end, committed) =>
+          decidedAction(stateRecovery)(end, committed)
+        }
       }
-    }
 
     val subscribeAndDecide = log.debug(s"subscribing to $subscription") *>
       stateConsumer.subscribe(subscription) *>
       log.debug("awaiting assignment") *>
       waitForAssignment.flatMap(a => log.debug(s"received assignment $a") *> decide(a).tap(d => log.debug(s"decided to $d")))
 
-    val initialize = subscribeAndDecide.flatMap {
-      case Initialize =>
-        log.info(s"consumer group $stateGroupId never consumed from $stateTopic") *>
-          stateProducer
-            .produce(stateTopic, key, initialState)
-            .provideLayer(registryLayer)
-            .tap(rmd => log.info(s"pushed initial state $initialState to $rmd"))
-            .unit
-      case Resume => log.info(s"consumer group $stateGroupId resuming consumption from $stateTopic")
-      case Retry | Fail => // TODO: fix handling of retry case
-        log.error(s"consumer group $stateGroupId is already at the end of $stateTopic, manual intervention required") *>
-          ZIO.fail(TamerError(s"Consumer group $stateGroupId stuck at end of stream"))
-    }
+    def initialize: RIO[Clock, Unit] =
+      subscribeAndDecide.flatMap {
+        case Initialize =>
+          log.info(s"consumer group $stateGroupId never consumed from $stateTopic") *>
+            stateProducer
+              .produce(stateTopic, key, initialState)
+              .provideLayer(registryLayer)
+              .tap(rmd => log.info(s"pushed initial state $initialState to $rmd"))
+              .unit
+        case Resume => log.info(s"consumer group $stateGroupId resuming consumption from $stateTopic")
+        case Recover(corrected) =>
+          log.error(s"incorrect consumer group's lag on $stateTopic. Group id: $stateGroupId. Unsubscribing the state consumer ...") *>
```

Comment:
Can these be `warning` instead of `error`? If I understand the code correctly `subscribeAndDecide` will short circuit the logic here so whatever being logged here is definitely a warning before taking a corrective action rather than logging and then decide to whether exit or continue.